uat/integrations/test_integrations.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import os
+
+import nbformat
+import pytest
+from nbclient.exceptions import CellExecutionError
+from nbconvert.preprocessors import ExecutePreprocessor
+
+# get the current working directory
+current_directory = os.getcwd()
+
+# change the working directory to the "examples" directory
+EXAMPLES_DIR = "examples"
+examples_directory = os.path.join(current_directory, EXAMPLES_DIR)
+os.chdir(examples_directory)
+
+KATIB_INTEGRATION = {"path": "katib-integration.ipynb", "id": "katib"}
+KFP_INTEGRATION = {"path": "kfp-integration.ipynb", "id": "kfp"}
+MINIO_INTEGRATION = {"path": "minio-integration.ipynb", "id": "minio"}
+MLFLOW_INTEGRATION = {"path": "mlflow-integration.ipynb", "id": "mlflow"}
+
+log = logging.getLogger(__name__)
+
+
+def format_error_message(traceback: list):
+    """Format error message."""
+    return "".join(traceback[-2:])
+
+
+@pytest.mark.ipynb
+@pytest.mark.parametrize(
+    # notebook - ipynb file to execute
+    "test_notebook",
+    [
+        KATIB_INTEGRATION["path"],
+        KFP_INTEGRATION["path"],
+        MINIO_INTEGRATION["path"],
+        MLFLOW_INTEGRATION["path"],
+    ],
+    ids=[
+        KATIB_INTEGRATION["id"],
+        KFP_INTEGRATION["id"],
+        MINIO_INTEGRATION["id"],
+        MLFLOW_INTEGRATION["id"],
+    ],
+)
+def test_integration(test_notebook):
```

Comment:
Is it possible for this to, at the end of each notebook test, save the final state of the notebook?  When debugging this feels really valuable.  I could then execute these tests, see a failure, and open the notebook to see exactly what happened plus have a good place to start iterating on a fix

Reply:
This is super interesting. Do you think it would ok UX-wise to have the tests alter the local directory, also given that it's a git repo? I mean, even if something doesn't go wrong, the notebook metadata will surely be different, which will result in a dirty repo every time we run the tests.

That said, I actually like this suggestion. The way I'm currently mounting the local directory to the Pod created by the K8s Job I'm pretty sure means that any changes inside the container affect the local directory. So it's a matter of configuring `nbconvert` to actually write the output back to the original notebook. I'm pretty hopeful that we'll be able to do that.

Reply:
So, I pushed a fix to save the output notebook back to the original one. However, there is still an issue there. For cells marked with `raises-exception` (i.e. cells performing checks) it works as expected; the notebook goes on executing and the errors in those cells are saved in its outputs. If an error is raised in one of the other cells, though, this won't work. In fact, it will look like this cell executed without any output. I could sort of patch that using the `allow_errors` flag. This is like adding the `raises-exception` tag everywhere and essentially means that should an error be raised in any cell, the execution will continue. This is not ideal IMO, since it can drown us in a bunch of errors, something that would not happen if executing the notebook manually (execution stops when a cell breaks). Although this would make the outputs be saved to the original notebook I don't really like it. Since an alternative solution doesn't look trivial, I'd say let's not worry too much about this for now. I'll open an issue for this once we have a list of next steps.

Reply:
ok opening an issue for this makes sense, it does go a bit out of scope

Funny enough, part of the inspiration for this in my head was Arrikto's kfp implementation.  Having each step driven by a notebook that I could open and debug when they failed always felt like a great feature :D

Reply:
I'll definitely get to the bottom of this, it will just take some more exploration to get it right, so it makes sense not to block on this for now, since this feels more like an enhancement.